pennylane/pulse/convenience_functions.py

```diff
@@ -115,3 +116,134 @@ def f(p, t):
         return _f(p, t)
 
     return f
+
+
+def pwc(timespan):
+    """Creates a function that is piecewise-constant in time.
+
+    Args:
+            timespan(Union[float, tuple(float, float)]: The timespan defining the region where the function is non-zero.
+              If an integer is provided, the timespan is defined as ``(0, timespan)``.
+
+    Returns:
+            func: a function that takes two arguments, an array of trainable parameters and a `float` defining the
+            time at which the function is evaluated. When called, the function uses the array of parameters to
+            create evenly sized bins within the ``timespan``, with each bin value set by an element of the array.
+            It then selects the value of the parameter array corresponding to the specified time, based on the
+            assigned binning.
+
+    **Example**
+
+    >>> timespan = (1, 3)
+    >>> f1 = pwc(timespan)
+
+    The resulting function ``f1`` has the call signature ``f1(params, t)``. If passed an array of parameters and
+    a time, it will assign the array as the constants in the piecewise function, and select the constant corresponding
+    to the specified time, based on the time interval defined by ``timespan``.
+
+    >>> params = [10, 11, 12, 13, 14]
+    >>> f1(params, 2)
+    Array(12, dtype=int32)
+
+    >>> f1(params, 2.1)  # same bin
+    Array(12, dtype=int32)
+
+    >>> f1(params, 2.5)  # next bin
+    Array(13, dtype=int32)
+    """
+    if not has_jax:
+        raise ImportError(
+            "Module jax is required for any pulse-related convenience function. "
+            "You can install jax via: pip install jax"
+        )
+
+    if isinstance(timespan, tuple):
+        t1, t2 = timespan
+    else:
+        t1 = 0
+        t2 = timespan
+
+    def func(params, t):
+        num_bins = len(params)
+        # include 0 as an additional option for function output
+        params = jnp.array(list(params) + [0])
```

Comment:
```suggestion
        params = jnp.concatenate([params, jnp.zeros(1)])
```
Creating a `list` is slowing it down
![image](https://user-images.githubusercontent.com/43949391/214641917-9a250e63-8a1c-407d-b002-4b1592eaf877.png)

Reply:
I see what you mean about outputting 0 out of bounds, but 0 is at least _sort of_ correct. 

If we remove this, since indexing can be negative, it doesn't raise an error when you request an out of bounds index on the lower end, it returns one of the parameters from the array (i.e. params[-2]). Which is more incorrect than returning 0. 

I don't know how to raise an error without an `if` clause that requires the index to be concrete. Is there a jitting-friendly way to say that if the index is negative, an error should be raised?

Reply:
I'll test the performance now

Reply:
fixed the issue 👍 thanks!